### ISSUE TITLE
nested recursive routes

### ISSUE THREAD

**Issue by mohammadne** [Milad-Akarie/auto_route_library#460]
what I have is that in example, I want to navigate from `MyBooksPage` to `BookDetailsPage` and vice versa.

my solution:

```dart
@MaterialAutoRouter(
  replaceInRouteName: 'Page,Route',
  routes: <AutoRoute>[
    AutoRoute(
      path: '/',
      page: HomePage,
      guards: [AuthGuard],
      usesTabsRouter: true,
      children: [
        AutoRoute(
          path: 'books',
          page: EmptyRouterPage,
          name: 'BooksTab',
          children: [
            AutoRoute(path: '', page: BookListPage),
            ..._homePageRoutes,
          ],
        ),
        AutoRoute(
          path: 'profile',
          name: 'ProfileTab',
          page: EmptyRouterPage,
          children: [
            AutoRoute(path: '', page: ProfilePage),
            ..._homePageRoutes,
          ],
        ),
        AutoRoute(
          path: 'settings',
          page: SettingsPage,
          name: 'SettingsTab',
        ),
      ],
    ),
    userDataRoutes,
    AutoRoute(path: '/login', page: LoginPage, fullscreenDialog: false),
    RedirectRoute(path: '*', redirectTo: '/'),
  ],
)
class $AppRouter {}

const List _homePageRoutes = <AutoRoute>[
  AutoRoute(path: 'books:id', page: BookDetailsPage),
  AutoRoute(path: 'myBooks', page: MyBooksPage),
];
```

error I will get :
```
There are more than one rout with the name [MyBooksRoute], route names must be unique!
Note: Unless specified, route name is generated from page name.
package:auto_test/router/router.dart:53:7
53 │ class $AppRouter {}
     │          ^^^^^^^^^^
```

I think another solution would be to use `AutoRouter` alongside `AutoTabsRouter` in `HomePage`.
but because it doesn't documented, I don't know how to use `AutoRouter` alongside `AutoTabsRouter` 
and put pages in the children of `HomePage` and create a `AutoRouter` scope for `HomePage` to navigate for it's children:
```dart
    AutoRoute(
      path: '/',
      page: HomePage,
      guards: [AuthGuard],
      usesTabsRouter: true,
      children: [
          // ...
       ]
```

in my case I want to navigate to `albumPage` from any tab, any nested pages like `playlistPage`, `artistPage`, so reasonable solution seems to put all pages (tabPages and others) in child routes of `HomePage` to have access to them instead of multiple duplication and put `albumPage` in the child of many routes, also second solution will solve recursive navigation issue.

**Comment by mohammadne** [comment]
also doing :
```dart
    AutoRoute(
      path: '/',
      page: EmptyRouterPage,
      guards: [AuthGuard],
      usesTabsRouter: true,
      children: [
        AutoRoute(path: '', page: HomePage),
        // ...
      ]
```

is useless because:
- bottom navigation will missed for nonTab children of `HomePage`
- type 'ParallelBranchEntry' is not a subtype of type 'StackRouter?' in type cast

**Comment by mohammadne** [comment]
the solution I guess that will solve it is to add a parameter to `AutoRoute` to add `AutoRouter` scope.

**Comment by mohammadne** [comment]
hey @Milad-Akarie 
is there any solution ?

**Comment by Milad-Akarie** [comment]
@mohammadne Sorry for the delay bro, I had in emergency at work yesterday. 
Generated route objects are merely names for predefined locations, the current setup will not allow you to use the same name for different locations but I was having some thoughts about making this a big leaner and allow for duplicate relative-location names as long as they link to the same page + have the same path. 
```
/first-scope/common-page
/second-scope/common-page
```

AutoRoute 2.0 is all about deep-location and deep-linking so I don't think it's a good idea to allow routes to be pushed anywhere regardless of their predefined locations.

**Comment by mohammadne** [comment]
thanks @Milad-Akarie , in my case, the first tab is my feed page and second one is search tab, in both tabs I want to navigate to `album_detail_page` and also there's a page named `albums_list_page` which I want to navigate to `album_detail_page`.
what do I should do here ?

also there are some other pages like artist and playlist, ... which are also the same.
for temp fix, I will duplicate these pages in all tabs and pages.

**Comment by Milad-Akarie** [comment]
Hey @mohammadne can't you assign these as root routes so you can push them from anywhere?

**Comment by mohammadne** [comment]
> Hey @mohammadne can't you assign these as root routes so you can push them from anywhere?

no, becase I`ll lose bottom_navigation after navigation.

**Comment by RicardoRB** [comment]
Seems similar to my question #468

**Comment by mohammadne** [comment]
hi @Milad-Akarie
is it possible to implement `deep-location` and `deep-linking` in coming days?

**Comment by mohammadne** [comment]
hi @Milad-Akarie 
thanks, this is now fixed in 2.0.2
